### ISSUE TITLE
ArrayInput and ReferenceInput render but generate an error -- with React 16.4.0

### ISSUE THREAD

**Issue by PhenixH** [marmelab/react-admin#1871]
**What you were expecting:**
ReferenceInput working with ArrayInput without having an error in the console.

**What happened instead:**
It's not exactly like #1806 : here the content is rendered but i still have the same error in the console : 

> Warning : Failed prop type: The prop `resource` is marked as required in `ReferenceInput`, but its value is `undefined`.

**Steps to reproduce:**
Use the code below
**Related code:**

* in my message.js

```
<ArrayInput source="payload.array"> 
				<SimpleFormIterator>
					<ReferenceInput label="product" source="productRef" reference="Products" perPage={10}  filterToQuery={searchFunct}>
						<AutocompleteInput optionText={optionRenderer} options={{
							filter: AutoComplete.caseInsensitiveFilter,
							maxSearchResults: 10
						}}/>
					</ReferenceInput>
				</SimpleFormIterator>
</ArrayInput>
```

**Other information:**

* Stack trace

```
    in ReferenceInput (created by TranslatedComponent(undefined))
    in TranslatedComponent(undefined) (created by ConnectedField)
    in ConnectedField (created by Connect(ConnectedField))
    in Connect(ConnectedField) (created by Field)
    in Field (created by FormField)
    in FormField (created by Connect(FormField))
    in Connect(FormField) (created by WithFormField)
    in WithFormField (at messages.js:105)
    in div (created by FormInput)
    in FormInput (created by WithStyles(FormInput))
    in WithStyles(FormInput) (created by SimpleFormIterator)
    in section (created by SimpleFormIterator)
    in li (created by SimpleFormIterator)
    in Transition (created by CSSTransition)
    in CSSTransition (created by SimpleFormIterator)
    in div (created by TransitionGroup)
    in TransitionGroup (created by SimpleFormIterator)
    in ul (created by SimpleFormIterator)
    in SimpleFormIterator (created by WithStyles(SimpleFormIterator))
    in WithStyles(SimpleFormIterator) (created by TranslatedComponent(WithStyles(SimpleFormIterator)))
    in TranslatedComponent(WithStyles(SimpleFormIterator)) (at messages.js:104)
    in Unknown (created by ConnectedFieldArray)
    in ConnectedFieldArray (created by Connect(ConnectedFieldArray))
    in Connect(ConnectedFieldArray) (created by FieldArray)
    in FieldArray (created by ArrayInput)
    in div (created by FormControl)
    in FormControl (created by WithStyles(FormControl))
    in WithStyles(FormControl) (created by ArrayInput)
    in ArrayInput (at messages.js:103)
    in div (created by FormInput)
    in FormInput (created by WithStyles(FormInput))
    in WithStyles(FormInput) (created by SimpleForm)
    in div (created by SimpleForm)
    in form (created by SimpleForm)
    in SimpleForm (created by WithStyles(SimpleForm))
    in WithStyles(SimpleForm) (created by Form(WithStyles(SimpleForm)))
    in Form(WithStyles(SimpleForm)) (created by Connect(Form(WithStyles(SimpleForm))))
    in Connect(Form(WithStyles(SimpleForm))) (created by ReduxForm)
    in ReduxForm (created by TranslatedComponent(undefined))
    in TranslatedComponent(undefined) (created by Connect(TranslatedComponent(undefined)))
    in Connect(TranslatedComponent(undefined)) (at messages.js:99)
    in div (created by Paper)
    in Paper (created by WithStyles(Paper))
    in WithStyles(Paper) (created by Card)
    in Card (created by WithStyles(Card))
    in WithStyles(Card) (created by CreateView)
    in div (created by CreateView)
    in CreateView (created by CreateController)
    in CreateController (created by TranslatedComponent(undefined))
    in TranslatedComponent(undefined) (created by Connect(TranslatedComponent(undefined)))
    in Connect(TranslatedComponent(undefined)) (created by Create)
    in Create (at messages.js:98)
    in PostCreate (created by WithPermissions)
    in WithPermissions (created by Connect(WithPermissions))
    in Connect(WithPermissions) (created by getContext(Connect(WithPermissions)))
    in getContext(Connect(WithPermissions)) (created by Route)
    in Route (created by Resource)
    in Switch (created by Resource)
    in Resource (created by Connect(Resource))
    in Connect(Resource) (at App.js:26)
    in Route (created by RoutesWithLayout)
    in Switch (created by RoutesWithLayout)
    in RoutesWithLayout (created by Route)
    in div (created by Layout)
    in main (created by Layout)
    in div (created by Layout)
    in div (created by Layout)
    in Layout (created by WithStyles(Layout))
    in WithStyles(Layout) (created by Connect(WithStyles(Layout)))
    in Connect(WithStyles(Layout)) (created by LayoutWithTheme)
    in MuiThemeProvider (created by LayoutWithTheme)
    in LayoutWithTheme (created by Route)
    in Route (created by CoreAdminRouter)
    in Switch (created by CoreAdminRouter)
    in div (created by CoreAdminRouter)
    in CoreAdminRouter (created by Connect(CoreAdminRouter))
    in Connect(CoreAdminRouter) (created by getContext(Connect(CoreAdminRouter)))
    in getContext(Connect(CoreAdminRouter)) (created by Route)
    in Route (created by CoreAdmin)
    in Switch (created by CoreAdmin)
    in Router (created by ConnectedRouter)
    in ConnectedRouter (created by CoreAdmin)
    in TranslationProvider (created by withContext(TranslationProvider))
    in withContext(TranslationProvider) (created by Connect(withContext(TranslationProvider)))
    in Connect(withContext(TranslationProvider)) (created by CoreAdmin)
    in Provider (created by CoreAdmin)
    in CoreAdmin (created by withContext(CoreAdmin))
    in withContext(CoreAdmin) (at App.js:25)
    in App (at index.js:7)`
```

**Environment**

* React-admin version: 2.0.2
* Last version that did not exhibit the issue (if applicable): none
* React version: 16.4.0
* Browser: Chrome
* Stack trace (in case of a JS error):

**Comment by fzaninotto** [comment]
Just out of curiosity: What's your use case for having an array of references?

**Comment by PhenixH** [comment]
One of my use cases is to provide an Order which give the choice to select products ( a huge database that is managed to by the API -> a Reference) and to specify for each product some options and customizations (such as the amount).

To sum up : for one order you can choose many products and each product can be combined with some options.

Do you see what i mean ?

**Comment by fzaninotto** [comment]
Ok, that makes sense, thanks. I still need to try and reproduce the issue. It would help if you could provide a codesandbox showcasing the error.

**Comment by PhenixH** [comment]
Yes, of course.
I will first recreate a clean environment and i will come back to you (while i am trying de figure it out my self ;) )

**Comment by PhenixH** [comment]
After my investigation i found that the error is generated when we click on "Add" button (generated by ArrayInput".

When i add the resource prop manually in SimpleFormIterator error disappear.
`<SimpleFormIterator resource="Orders">`

It seems that ArrayInput is not passing resource to SimpleFormIterator ( as it was with record before #1792 )

So my suggestion is to do a similar fix to #1792 .

What do you think ?

**Comment by PhenixH** [comment]
So i made a sandbox  : first i was not able to reproduce the bug, but i figured it out : it depends on react version.
It seems that with 16.3.1, there is no problem but with 16.4.0 it generate the bug.
Yes i know, i am playing with fire by using a react version that is not supported yet by React Admin. I am preparing the future 😄  

I will change the title of the issue.

Use this sandbox on Post Create, Add a Co-Writer (arrayofCowriter), see the error in log
https://codesandbox.io/s/mmrzy16x4p
